### ISSUE TITLE
[BUG] FlatteningSerializer not work correctly for Map/List with final modifier

### ISSUE THREAD

**Issue by weidongxu-microsoft** [Azure/azure-sdk-for-java#20150]
**Describe the bug**

It breaks the "dot" in Map key, result in e.g. `{"operation":"Replace","properties":{"tags":{"tag":{"1":"value.1"}}}}`

Sample case https://github.com/Azure/azure-sdk-for-java/pull/20149

Code on https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningSerializer.java#L106-L122

It happens to track1, reported by Databricks. Checked on track2, still reproducible (though it is now limited to classes annotated with `@JsonFlatten`).

I am not able to find good solution for this, as the `Map` or `List` might not even be modifiable.

***Exception or Stack Trace***
Add the exception log and stack trace if available

**To Reproduce**

Sample case https://github.com/Azure/azure-sdk-for-java/pull/20149

***Code Snippet***
Add the code snippet that causes the issue.

Sample case https://github.com/Azure/azure-sdk-for-java/pull/20149

**Expected behavior**

**Screenshots**

**Setup (please complete the following information):**
 - IDE : [e.g. IntelliJ]
 - Version of the Library used: azure-core 1.14.1

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by weidongxu-microsoft** [comment]
@anuchandy @jianghaolu for awareness.

**Comment by weidongxu-microsoft** [comment]
Another case found:

```
@JsonFlatten ClassA
-- propertyA: final ClassB
---- propertyB: Map<>
```

The escaping of Map get interrupted by the "final" ClassB via https://github.com/Azure/azure-sdk-for-java/blob/azure-core_1.6.0/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningSerializer.java#L121-L123

It could be solved by the "shallow" JsonFlatten (however I am not sure what might be the impact of changing behavior from "deep" to "shallow").

**Comment by samvaity** [comment]
Fixed with ##22591 to allow flattening of static and final variables.
